### PR TITLE
Add Rust specific build info to metadata

### PR DESCRIPTION
**Pull request by HCastano**
Will close the metadata part of #525.

This is what the new metadata looks like:

```json
  "source": {
    "hash": "0xad8e1a3d7d1c44b4c7b3f715765ab8003e15aa2d8d1146a64cbc9b8f7e0a8df9",
    "language": "ink! 4.0.0-alpha.3",
    "compiler": "rustc 1.64.0",
    "build_info": {
      "build_mode": "Release",
      "cargo_contract_version": "2.0.0-alpha.4",
      "rustc_version": "stable-x86_64-unknown-linux-gnu",
      "wasm_opt_settings": {
        "keep_debug_symbols": false,
        "optimization_passes": "Z"
      }
    }
```

The stuff related to the `verify` command will be done in a follow-up.

Note: I've ended up removing the `wasm-opt` version from the `build_info` for now since there's
no easy way to get it from the new `wasm-opt` library. 

I'm going to assume that two installations of `cargo-contract` contain the same `wasm-opt` 
library (this only holds if both were installed with `--locked`). We should add it back in the future
in order to give users a better indication of what went wrong during the `verify` phase.

### CODE REVIEW

**Comment by athei** [comment]
I don't think we should add the wasm-opt version. We should have a useable `wasm-opt` version as crate as early as next week. Then the version is described by the manifest alread.

**Comment by HCastano** [comment]
> I don't think we should add the wasm-opt version. We should have a useable `wasm-opt` version as crate as early as next week. Then the version is described by the manifest alread.

Sure, but until that happens this is still needed. We can get rid of that field once that crate is published.

File: crates/cargo-contract/src/cmd/metadata.rs [hunk 1]
```diff
@@ -73,6 +75,42 @@ struct ExtendedMetadataResult {
...
 }
 
+/// Information about the settings used to build a particular ink! contract.
+///
+/// Note that this should be an optional part of the metadata since it may not necessarily
+/// translate to other languages (e.g ask!, Solidity).
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+pub struct BuildInfo {
+    /// The stable version of `rustc` used to build the contract.
+    pub rustc_version: Version,
```
Comment (ascjones) [inline crates/cargo-contract/src/cmd/metadata.rs:85]: Can/should we also add the toolchain which contains the architecture? e.g. `stable-x86_64-unknown-linux-gnu`. I remember there being some discussions about this affecting the wasm bytecode output, can't remember what the conclusions were.

File: crates/cargo-contract/src/cmd/build/mod.rs [hunk 2]
```diff
@@ -650,7 +655,29 @@ pub(crate) fn execute(args: ExecuteArgs) -> Result<BuildResult> {
             &crate_metadata.contract_artifact_name,
         )?;
 
-        Ok(optimization_result)
+        let cargo_contract_version =
+            if let Ok(version) = Version::parse(env!("CARGO_PKG_VERSION")) {
```
Comment (ascjones) [inline crates/cargo-contract/src/cmd/build/mod.rs:659]: AFAIU `CARGO_PKG_VERSION` is only available at compile time (and also during `cargo run`). So would need to be captured in a `const` e.g. `const VERSION: &'static str = env!("CARGO_PKG_VERSION");`

File: crates/cargo-contract/src/cmd/build/mod.rs [hunk 3]
```diff
@@ -650,7 +658,26 @@ pub(crate) fn execute(args: ExecuteArgs) -> Result<BuildResult> {
...
+            version
+        } else {
+            anyhow::bail!(
+                "Unable to parse version number for the currently running \
+                    `cargo-contract` binary."
+            );
+        };
+
+        let build_info = BuildInfo {
+            rustc_version: crate::util::rustc_toolchain()?,
```
Comment (ascjones) [inline crates/cargo-contract/src/cmd/build/mod.rs:671]: This field should be renamed `rust_toolchain` or so, as the actual version of `rustc`is in the `compiler` field/

File: crates/metadata/src/lib.rs [hunk 4]
```diff
@@ -137,6 +139,11 @@ pub struct Source {
     /// with the metadata.
     #[serde(skip_serializing_if = "Option::is_none")]
     pub wasm: Option<SourceWasm>,
+    /// Extra information about the environment in which the contract was built.
+    ///
+    /// Useful for producing deterministic builds.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub build_info: Option<Map<String, Value>>,
```
Comment (ascjones) [inline crates/metadata/src/lib.rs:146]: Curious why we convert to loosely typed json `Map` here? Is that for alt. compilers e.g. `solang` to provide arbitrary data here?

**Review by ascjones — APPROVED** [review]

File: crates/metadata/src/lib.rs (hunk 4 above)
Comment (HCastano) [reply crates/metadata/src/lib.rs:146]: Yep, other compilers may want to include other info here. It's also useful for us if we decide that we need to add/remove info later on

**Pull request merged**